### DISCUSSION TITLE
Ability to Edit Workflow YAML from GUI

### DISCUSSION THREAD

**Discussion by EatMoreChicken** [langgenius/dify#3330] [Suggestion]
### Self Checks

- [X] I have searched for existing issues [search for existing issues](https://github.com/langgenius/dify/issues), including closed ones.
- [X] I confirm that I am using English to submit this report (我已阅读并同意 [Language Policy](https://github.com/langgenius/dify/issues/1542)).
- [X] Pleas do not modify this template :) and fill in all the required fields.

### 1. Is this request related to a challenge you're experiencing?

When I'm working on building a workflow, I often times find myself needing to add multiple similar items and I have to manually click through and build them. For example, adding classes to the question classifier when there are many to add.

### 2. Describe the feature you'd like to see

I would like to see a method to edit the workflow's YAML from the GUI. This was we can mass add elements to the workflow without needing to click through. For example, with the question classifier, I can can quickly use regex or other transformation method to build out the different classes and get them added by copy and pasting the elements. If there an error with the YAML parsing, it can prevent the user from saving it somehow.

### 3. How will this feature improve your workflow or experience?

This will make it much easier to bulk add similar blocks or elements within blocks.

### 4. Additional context or comments

There might need to be a way to auto detect these additions and generate IDs were needed.

### 5. Can you help us with this feature?

- [ ] I am interested in contributing to this feature.

**Comment by crazywoola** [comment]
But the dify itself is a GUI tool for yaml file right? Why do you still need a GUI tool?
Personally, I don't think GUI is a good way for bulk editing stuffs like that.

**Reply by EatMoreChicken** [reply]
For sure. The GUI tool is very useful for visualizing the flows, but having a way to edit the YAML directly for workflows when needed will help speed up bulk additions or edits. I've seen this implemented in other tools and wanted to point it out as feature that would be useful.

This is a completely different example, but Splunk lets users build dashboard using a GUI tool, but there is also an option to see the raw XML source of the dashboard so you can edit elements in there. An example with dashboards is to be able to bulk edit panel names or make copied of existing panels without having to manually add and fill in information for each new panel.

**Reply by EatMoreChicken** [reply]
Maybe another example in terms of Dify would be that it unlocks the ability to copy and paste in someone else's block directly into the YAML as a way to import.

**Comment by perzeuss** [comment]
I think editing the whole YAML is not a good solution. You might break things if you do unintended modification. It might be a solution for a few people but it does not resolve the problem behind your pain with the editor.

What we should do instead is improving the user experience of the workflow editor. Here are some examples:

- Allow to select multiple nodes to move them at the same time
- Allow to copy & paste nodes by selecting them and using keyboard shortcuts
- Allow to duplicate nodes by selecting one and using a keyboard shortcut
- Implement a bulk-edit mode for classifier to allow adding multiple classes in one text field using a line break as separator
- Add an way to edit multiple nodes at once, if all selected nodes are the same type (e.g. Render the LLM node UI if I select 5 LLM nodes. If I make changes, e.g. change the model, the change will be applied to all selected nodes) 
  - There could pop up a button in the corner of the editor If I have multiple nodes selected and all are of the same type. This button could open the node in a "shared edit mode" which allows to sync changes to all selected nodes  
- Add a menu for bulk tasks when multiple nodes are selected (even if they are not the same type of node), for:
  - Editing the title of all selected nodes 
  - Editing the description of all selected nodes
  - Deleting the nodes
  - Duplicating the nodes
- Add a "Variable Creator" node, which allows to define multiple variables to re-use values in the workflow (e.g. a system instruction that I can link in all LLM nodes, or a HTTP header which I need in multiple HTTP Request nodes)
  - We could do that with a code block but this is very slow
- Allow to re-order classes in the Question Classifier node
- Allow to group nodes, colorize groups and move them around (large workflows are really a pain to maintain and can become really messy)
